### PR TITLE
feat(filter-field): Interaction-state

### PR DESCRIPTION
**Pull request by rowa-audil**
### <strong>Pull Request</strong>

The interaction state represents whether the user currently is interacting with the filter-field or not.
_Adds a Member `interactionState` and an Output `interactionStateChange`_

- Feel free to give feedback on the naming!

#### Type of PR

Feature (non-breaking change which adds functionality)

#### Checklist

- [x] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 1]
```diff
@@ -579,6 +585,20 @@ export class DtFilterField<T = any>
...
+        this.interactionState = true;
+        this.interactionStateChange.emit(true);
+      });
+
+    this._autocomplete.closed
+      .pipe(startWith(null), takeUntil(this._destroy$))
+      .subscribe(() => {
+        this.interactionState = false;
+        this.interactionStateChange.emit(false);
+      });
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field.ts:601]: I don't think this covers the whole range of the interaction. The autocomplete can be closed in this case if the current filter is a range (or multi-select later on when #1836 is done). 

Additionally, a autocomplete does not have to be open while interacting (free text filter).

**Review by tomheller — CHANGES_REQUESTED** [review]

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 2]
```diff
@@ -510,10 +519,38 @@ export class DtFilterField<T = any>
...
+        if (this._filterfieldRange.isOpen || this._autocomplete.isOpen) {
+          this.interactionStateChange.emit(true);
+        } else {
+          this.interactionStateChange.emit(false);
+        }
+      });
+
+    this._filterfieldRange.closed.subscribe(() => {
+      // might need to check if the input is focused again
+      this._isFocusedSubject.next(null);
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field.ts:552]: Please clean up the subscription as well. What about the range closing?

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 3]
```diff
@@ -510,10 +519,38 @@ export class DtFilterField<T = any>
...
+        if (isFocused) {
+          this.interactionStateChange.emit(true);
+          return;
+        }
+        // If the range is open then we are still in the filter-field context so the interaction state is active
+        if (this._filterfieldRange.isOpen || this._autocomplete.isOpen) {
+          this.interactionStateChange.emit(true);
+        } else {
+          this.interactionStateChange.emit(false);
+        }
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field.ts:547]: nitpick
```suggestion
       this.interactionStateChange.emit(this._filterfieldRange.isOpen || this._autocomplete.isOpen)
```

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 4]
```diff
@@ -339,6 +342,9 @@ export class DtFilterField<T = any>
     delay(0),
   );
 
+  /** Whether the filter-field is being interacted with */
+  interactionState: boolean;
+
```
Comment (tomheller) [inline libs/barista-components/filter-field/src/filter-field.ts:347]: You are never setting the interactionState. 
It might be a good idea to create a method that will do a couple of things: 
1. set the interactionState member
2. Checks if the interactionState has changed in regards to the previous value
3. If 2., emit the interactionStateChange.

**Review by tomheller — CHANGES_REQUESTED** [review]
I don't believe the functionality already works as intended. You are on a good track though. 
Some tests that check interactionState and listen to the event might help you here. 

Please double check that you do not emit interactionStateChange if the value actually did not change (if it switches from true to true).

**Comment by rowa-audil** [comment]
@tomheller 
When the consumer uses free text and clicks outside the filter-field the interaction state will stay on true. My case is, the interaction hasn't finished yet, even though the focus is somewhere else.

I've looked over the use-cases where the interaction state changes.
Am I missing something?

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 5]
```diff
@@ -445,6 +451,10 @@ export class DtFilterField<T = any>
   /** Whether the filter field or one of it's child elements is focused. */
   private _isFocused = false;
```
Comment (rowa-audil) [inline libs/barista-components/filter-field/src/filter-field.ts:453]: I missed the JSdoc. It will be added with the squash and rebase.

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 6]
```diff
@@ -445,6 +451,10 @@ export class DtFilterField<T = any>
...
 
+  private _mergedAutocompleteRangeOpened: Observable<unknown>;
+
```
Comment (rowa-audil) [inline libs/barista-components/filter-field/src/filter-field.ts:455]: I missed the JSdoc. It will be added with the squash and rebase.

File: apps/dev/src/filter-field/filter-field-demo.component.ts [hunk 7]
```diff
@@ -152,6 +152,10 @@ export class FilterFieldDemo implements AfterViewInit, OnDestroy {
         this._firstTag = tags[0];
       }
     });
+
```
Comment (rowa-audil) [inline apps/dev/src/filter-field/filter-field-demo.component.ts:155]: Will be removed with rebase.

**Review by rowa-audil — COMMENTED** [review]
Some things to be removed after rebase/squash

File: apps/dev/src/filter-field/filter-field-demo.component.ts [hunk 8]
```diff
@@ -152,6 +152,10 @@ export class FilterFieldDemo implements AfterViewInit, OnDestroy {
...
+    this.filterField.interactionStateChange.subscribe((isInteracted) => {
```
Comment (rowa-audil) [inline apps/dev/src/filter-field/filter-field-demo.component.ts:156]: Will be removed with rebase.

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 9]
```diff
@@ -445,6 +451,10 @@ export class DtFilterField<T = any>
   /** Whether the filter field or one of it's child elements is focused. */
   private _isFocused = false;
 
+  private _mergedAutocompleteRangeOpened: Observable<unknown>;
```
Comment (rowa-audil) [inline libs/barista-components/filter-field/src/filter-field.ts:454]: JsDoc will be added with rebase.

File: libs/barista-components/filter-field/src/filter-field.ts [hunk 10]
```diff
@@ -445,6 +451,10 @@ export class DtFilterField<T = any>
...
+
+  private _mergedAutocompleteRangeClosed: Observable<unknown>;
```
Comment (rowa-audil) [inline libs/barista-components/filter-field/src/filter-field.ts:456]: JsDoc will be added with rebase.

File: libs/barista-components/filter-field/src/filter-field.ts (hunk 9 above)
Comment (rowa-audil) [inline libs/barista-components/filter-field/src/filter-field.ts:454]: Missed the JsDoc. Will be added with rebase.

File: apps/dev/src/filter-field/filter-field-demo.component.ts (hunk 8 above)
Comment (rowa-audil) [inline apps/dev/src/filter-field/filter-field-demo.component.ts:156]: Will be removed with rebase.

Comment (rowa-audil) [reply apps/dev/src/filter-field/filter-field-demo.component.ts:156]: We could leave that actually

Comment (tomheller) [reply apps/dev/src/filter-field/filter-field-demo.component.ts:156]: we can leave that in the devapp, you are right

File: libs/examples/src/filter-field/filter-field-unique-example/filter-field-unique-example.ts [hunk 11]
```diff
@@ -56,4 +57,16 @@ export class DtExampleFilterFieldUnique {
   };
 
   _dataSource = new DtFilterFieldDefaultDataSource(this.DATA);
+
+  @ViewChild(DtFilterField) filterField: DtFilterField;
+
+  ngAfterViewInit(): void {
+    this.filterField.interactionStateChange.subscribe((state) => {
```
Comment (tomheller) [inline libs/examples/src/filter-field/filter-field-unique-example/filter-field-unique-example.ts:64]: Please clear this subscription as well.

**Review by tomheller — CHANGES_REQUESTED** [review]
I've tried this one in the dev app and it seems as if it never fires the `interactionState: false` when the filter field blurs with keyboard.

**Review by tomheller — APPROVED** [review]
Looks good, maybe reword the commit message to the following: 

```
feat(filter-field): Added interaction state handling and notifications.

This should be helpful to the consumer to determine when it is safe to update the datasource of the filter field, without breaking the experience of the user. 

Fixes #1321
```

**Review by ffriedl89 — APPROVED** [review]
